### PR TITLE
Fixes #1474 - added  pre-commit linter

### LINKED ISSUE #1474: Document recommended eslint/prettier workflow
Support for prettier landed in #1468.

We should determine a simple workflow, and document it so we're all on the same page.

Strawman proposal:

 1.  write some code
 2.   run npm run lint
 3.   run npm run fix (if there were simple style errors...)
 4.   commit, make PR, etc.

### PR DESCRIPTION
**Pull request by magsout**
I just added the pre-commit part https://github.com/webcompat/webcompat.com/issues/1474#issuecomment-291633586

r? @zoepage

### CODE REVIEW

**Comment by zoepage** [comment]
@magsout Sorry for the late review. Had issues with my setup. 

I get two errors. Could you take a look please? :)
```
/webcompat/webcompat.com/webcompat/static/js/lib/labels.js
  159:59  error  Follow `prettier` formatting (expected '\n' but found ' ')  prettier/prettier
/webcompat/webcompat.com/webcompat/static/js/lib/models/query-params.js
  278:28  error  Follow `prettier` formatting (expected '\n' but found ' ')  prettier/prettier

✖ 2 problems (2 errors, 0 warnings)
```

Would you also like to add the information about the pre-linter to that PR? Would be great. Otherwise I'm happy to file a follow up issue and add it.
https://github.com/webcompat/webcompat.com/issues/1474#issuecomment-297182085

**Comment by magsout** [comment]
>I get two errors. Could you take a look please? :)

Yeah, I had exactly these errors. I think we should blocked version to avoid this kind of error. If you check the version of `prettier` in node_modules it is different than 1.1.0..

**Comment by magsout** [comment]
>Would you also like to add the information about the pre-linter to that PR? Would be great. Otherwise I'm happy to file a follow up issue and add it.
#1474 (comment)

@zoepage oups, I I have forgotten, I'll do, tahnks ;)

**Comment by zoepage** [comment]
@magsout I know, you are super busy right now (FAMILY always comes FIRST 🎉)
Is there anything I could help with? (If not, don't worry.)

**Comment by magsout** [comment]
>@magsout I know, you are super busy right now (FAMILY always comes FIRST 🎉)
Is there anything I could help with? (If not, don't worry.)

@zoepage oups, sorry, I'm on it. I'll take care of it Friday night...

**Comment by magsout** [comment]
@zoepage sorry for the delay..

I have sent a commit for the documentation. Tell me what do you think ?

File: docs/pr-coding-guidelines.md [hunk 1]
```diff
@@ -144,10 +144,18 @@ We do not use frameworks. However we use libraries, such suitcss-components-grid
...
 
+#### Linting
+
+For the consistency of our codebase, we use eslint with these rules: https://github.com/webcompat/webcompat.com/blob/master/.eslintrc. We also use Prettier for formatting our code. We have set up two recipes to check and fixe the codebase.
+
+`npm run lint` check all JavaScript files and display the resulting errors. If you get errors, you have two possibilities. You can correct yourself or you can run `npm run fix` which use Prettier to format the code. 
+
+In order to avoid errors during a Pull Request, `npm run lint` will be executed before each commit.
+
 @@something to write by miketaylr@@
```
Comment (zoepage) [inline docs/pr-coding-guidelines.md]: @miketaylr ? :D

File: docs/pr-coding-guidelines.md [hunk 2]
```diff
@@ -144,10 +144,18 @@ We do not use frameworks. However we use libraries, such suitcss-components-grid
...
 ### Javascript
 
-The js folder contains two subfolders: `lib` contains all project source files and `vendor` contains all third party libraries. The files out of the two sub folders contain the compiled source code.
+The js folder contains two sub folders: `lib` contains all project source files and `vendor` contains all third party libraries. The files out of the two sub folders contain the compiled source code.
 
 > Note: All code changes should be made to the files in `lib`
 
+#### Linting
+
+For the consistency of our codebase, we use eslint with these rules: https://github.com/webcompat/webcompat.com/blob/master/.eslintrc. We also use Prettier for formatting our code. We have set up two recipes to check and fixe the codebase.
```
Comment (zoepage) [inline docs/pr-coding-guidelines.md]: It's a great start @magsout! :)
I'd suggest to add a few more links to the tools and docs as following:

```
To make it easier for everyone to contribute, we use [eslint](http://eslint.org/) to keep the code base steady. You can view the rules in our [ESLint config](https://github.com/webcompat/webcompat.com/blob/master/.eslintrc#L34). For further explanation of the rules, please have a look at the [ESLint documentation](http://eslint.org/docs/rules/). 

For the linting process, we use [Prettier](http://jlongster.com/A-Prettier-Formatter). 
```

I'm not 100% sure what `recipes` means in this context. Could you explain it to me please? :)

File: docs/pr-coding-guidelines.md [hunk 3]
```diff
@@ -144,10 +144,18 @@ We do not use frameworks. However we use libraries, such suitcss-components-grid
...
+
+`npm run lint` check all JavaScript files and display the resulting errors. If you get errors, you have two possibilities. You can correct yourself or you can run `npm run fix` which use Prettier to format the code.
```
Comment (zoepage) [inline docs/pr-coding-guidelines.md]: Suggesting a little bit different phrasing. Feel free to change it as you like or dismiss, if you feel its not fitting. 

```
`npm run lint` checks all relevant JavaScript files and displays, if something needs to be fixed. 

If you get an error displayed, there are two ways to fix it. 
1. You can run `npm run fix` automatically, which is great for small issues like missing spaces or lines in various files. 
2. You can correct it manually as every error message includes the file and line of the error as well as the rule which was violated.
```

File: docs/pr-coding-guidelines.md [hunk 4]
```diff
@@ -144,10 +144,18 @@ We do not use frameworks. However we use libraries, such suitcss-components-grid
...
 
 > Note: All code changes should be made to the files in `lib`
 
+#### Linting
+
+For the consistency of our codebase, we use eslint with these rules: https://github.com/webcompat/webcompat.com/blob/master/.eslintrc. We also use Prettier for formatting our code. We have set up two recipes to check and fixe the codebase.
+
+`npm run lint` check all JavaScript files and display the resulting errors. If you get errors, you have two possibilities. You can correct yourself or you can run `npm run fix` which use Prettier to format the code. 
+
+In order to avoid errors during a Pull Request, `npm run lint` will be executed before each commit.
```
Comment (zoepage) [inline docs/pr-coding-guidelines.md]: Do you wanna pull this up between the first 2 paragraphs of `Linting`? Before the `npm run lint` paragraph?

File: package.json [hunk 5]
```diff
@@ -32,17 +36,17 @@
...
     "postcss": "^5.2.14",
     "postcss-browser-reporter": "^0.4.0",
     "postcss-cssnext": "^2.9.0",
     "postcss-import": "^8.0.2",
     "postcss-reporter": "^3.0.0",
     "postcss-url": "^5.1.1",
-    "prettier": "^1.1.0",
-    "suitcss-utils-align": "^0.2.0",
-    "suitcss-utils-display": "^0.4.0"
+    "prettier": "^1.1.0"
```
Comment (magsout) [inline package.json]: good idea

File: docs/pr-coding-guidelines.md (hunk 2 above)
Comment (magsout) [inline docs/pr-coding-guidelines.md]: recipes for me, it's all the scripts in package.json as `npm run *`

Comment (zoepage) [inline docs/pr-coding-guidelines.md]: Should we add an explanation somewhere? :)

**Comment by magsout** [comment]
Done @zoepage

**Comment by magsout** [comment]
let me know if it's ok. I'll rebase my commits

**Comment by zoepage** [comment]
@magsout I'll try to check ASAP. Might take 1-2 days. Sorry!

File: docs/pr-coding-guidelines.md [hunk 6]
```diff
@@ -150,12 +150,19 @@ The js folder contains two sub folders: `lib` contains all project source files
...
-`npm run lint` check all JavaScript files and display the resulting errors. If you get errors, you have two possibilities. You can correct yourself or you can run `npm run fix` which use Prettier to format the code. 
+For the linting process, we use [Prettier](http://jlongster.com/A-Prettier-Formatter).
 
 In order to avoid errors during a Pull Request, `npm run lint` will be executed before each commit.
 
+`npm run lint` checks all relevant JavaScript files and displays, if something needs to be fixed.
+
+If you get an error displayed, there are two ways to fix it.
+1. You can run `npm run fix` automatically, which is great for small issues like missing spaces or lines in various files.
+2. You can correct it manually as every error message includes the file and line of the error as well as the rule which was violated.
```
Comment (zoepage) [inline docs/pr-coding-guidelines.md]: "which was violated."-> "which was violated will be displayed."

(Sorry for probably suggesting a wrong fix)

**Comment by zoepage** [comment]
Hey @magsout!
Found one tiny thing (I'm sure, I didn't see it before or suggested a wrong fix).

After that, feel free to rebase and merge! Great job! 🎉

File: package.json [hunk 7]
```diff
@@ -57,7 +61,11 @@
     "virtualenv": "pip install virtualenv && virtualenv env && source env/bin/activate || . env/bin/activate && npm run pip",
     "pip": "pip install -r config/requirements.txt",
     "config": "cp config/secrets.py.example config/secrets.py",
-    "project-update": "pip install --upgrade pip && npm update"
+    "project-update": "pip install --upgrade pip && npm update",
+    "precommit": "lint-staged"
+  },
+  "lint-staged": {
+    "{webcompat/static/js/lib, tests, grunt-tasks}/**/*.js": "lint"
```
Comment (zoepage) [inline package.json]: The task is not fired up as the path seems not fitting, which is weird.
Let's make things easier and just switch from `"{webcompat/static/js/lib, tests, grunt-tasks}/**/*.js": "lint"` to `"*.js": "lint"`

So we are always on the safe side?

**Comment by magsout** [comment]
thanks a lot @zoepage for your review.  <3

**Pull request merged**